core/src/consensus.rs

```diff
@@ -864,16 +887,43 @@ impl Tower {
         );
 
         if let Some(last_voted_slot) = self.last_voted_slot() {
-            if slot_history.check(last_voted_slot) == Check::TooOld {
-                // We could try hard to anchor with other older votes, but opt to simplify the
-                // following logic
-                return Err(TowerError::TooOldTower(
+            if tower_root <= replayed_root {
+                // Normally, we goes into this clause with possible help of
+                // reconcile_blockstore_roots_with_tower()
+                if slot_history.check(last_voted_slot) == Check::TooOld {
+                    // We could try hard to anchor with other older votes, but opt to simplify the
+                    // following logic
+                    return Err(TowerError::TooOldTower(
+                        last_voted_slot,
+                        slot_history.oldest(),
+                    ));
+                }
+                self.adjust_lockouts_with_slot_history(slot_history)?;
+                self.initialize_root(replayed_root);
+            } else {
+                // This should never occur under normal operation.
+                // If this validator has leader slots while voting suspended this way,
+                // suspended_decision_due_to_major_unsynced_ledger() will be also touched.
+                let message = format!(
+                    "For some reason, we're REPROCESSING already voted and ROOTED slots; \
+                     VOTING will be SUSPENDED UNTIL {}!",
                     last_voted_slot,
-                    slot_history.oldest(),
-                ));
+                );
+                error!("{}", message);
+                datapoint_error!("tower_error", ("error", message, String));
+                // pass-through adjust_lockouts_with_slot_history just to sanitize without
+                // any adjustment using a synthesized SlotHistory.
+                let mut warped_slot_history = (*slot_history).clone();
+                warped_slot_history.add(tower_root);
```

Comment:
Follow-up to this: https://github.com/solana-labs/solana/pull/12671/files#r513942038, for maintainability, could we add a 
comment on why this `warped_slot_history.add(tower_root); ` is necessary and sufficient for correctness (I think to find an anchored slot)?

Reply:
added a comment: 2aba6a13eefab135acf6337d6d030b6918cf7293